nifty-ssl/src/main/java/com/facebook/nifty/ssl/PollingMultiFileWatcher.java

```diff
@@ -0,0 +1,275 @@
+     * called.
+     *
+     * @param initialDelay how long to wait until the first scan of the files.
+     * @param interval how often to rescan the files.
+     * @param timeUnit time unit for {@code initialDelay} and {@code interval}.
+     */
+    public PollingMultiFileWatcher(long initialDelay, long interval, TimeUnit timeUnit) {
+        this.initialDelay = initialDelay;
+        this.interval = interval;
+        this.timeUnit = timeUnit;
+        this.executorService = null;
+        this.future = null;
+        this.metadataCacheRef = new AtomicReference<>(ImmutableMap.of());
+        watchedFiles = ImmutableSet.of();
+        callback = null;
+    }
+
+    /**
+     * Starts polling the watched files for changes.
+     */
+    public void start(Set<File> files, Consumer<Set<File>> callback) {
+        if (isStarted()) {
+            throw new IllegalStateException("start() should not be called more than once");
+        }
+        Preconditions.checkArgument(files.size() > 0, "must specify at least 1 file to watch for changes");
+        this.callback = requireNonNull(callback);
+        watchedFiles = ImmutableSet.copyOf(files);
+        executorService = MoreExecutors.listeningDecorator(Executors.newScheduledThreadPool(1));
+        future = executorService.scheduleAtFixedRate(
+            this::scanFilesForChanges,
+            initialDelay,
+            interval,
+            timeUnit);
+    }
+
+    /**
+     * @return true if the polling thread has been started and shutdown has not been called.
+     */
+    public boolean isStarted() {
+        return executorService != null;
+    }
+
+    /**
+     * Stops polling the files for changes. Should be called during server shutdown or when this watcher is no
+     * longer needed to make sure the background thread is stopped.
+     */
+    public void shutdown() {
+        shutdown(false);
+    }
+
+    private void shutdown(boolean fromFinalizer) {
+        if (isStarted()) {
+            if (fromFinalizer) {
+                // Effective Java Item 7 - warn if a "safety net" finalizer actually did any work.
+                log.warn("%s garbage-collected but shutdown() was never called. " +
+                    "Don't rely on finalizers to clean up background threads.",
+                    this.getClass().getSimpleName());
+            }
+            future.cancel(true);
+            executorService.shutdown();
+            future = null;
+            executorService = null;
+            watchedFiles = ImmutableSet.of();
+            callback = null;
+            metadataCacheRef.set(ImmutableMap.of());
+        }
+    }
+
+    @Override protected void finalize() {
+        // Implement finalize() so we'll try to stop the background thread if caller forgets to do it.
+        // JVM provides no guarantees about when (or even if) the finalizer will run, so don't rely on it.
+        shutdown(true);
```

Comment:
log a serious error that shutdown wasn't called, so as not to (inconsistently) cover for mistakes.